Move ClientHello padding to the encoding. (#443)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

reference an extension in ClientHelloOuter.

When sending ClientHello, the client first computes ClientHelloInner, including
any PSK binders. It then computes a new value, the EncodedClientHelloInner,
which is the following structure:

~~~
    struct {
        ClientHello client_hello;

>         opaque padding[N];

Reviewer: nit: `uint8 zeros[length_of_padding];` aligns a little nicer with [8446](https://datatracker.ietf.org/doc/html/rfc8446#section-5.2).

Author: Done. It does make it no longer align with the requested variable name change in {{padding}}, but I agree with you aligning with 8446 makes sense.

>         opaque padding<0..2^16-1>;

Reviewer 2: This would be easier if this wasn't length-prefixed.  Every padding scheme that ever used length has caused a lot of pain.  I realize that this isn't necessarily security-critical, but just extending with zeros should be easier to implement AND you don't have two extra bytes of overhead.

Author: Works for me. Since every TLS structure is self-delimiting, we can just append stuff. It shifts some work from presentation language to prose, but I agree that it's easier to implement.

Author: Done. I kinda cheated a bit around the presentation language, but hopefully that's clear enough?

    } EncodedClientHelloInner;
~~~

The `client_hello` field is computed by first making a copy of

> ClientHelloInner with the `legacy_session_id` field with an empty string.

Reviewer 3: ```suggestion
ClientHelloInner and sets the `legacy_session_id` field to an empty string.
```

Author: Done, though s/sets/setting/

> Note this field uses the ClientHello structure, defined in Section 4.1.2
> of {{RFC8446}} which does not include the Handshake structure's four byte

Reviewer 2: ```suggestion
Note this field uses the ClientHello structure, defined in {{Section 4.1.2
of RFC8446}} which does not include the Handshake structure's four byte
```

Author: YOU CAN DO THAT?!?!!? Wow, I need to go update all my documents now.

> header.

Reviewer 3: ```suggestion
header. The `padding` field MUST be all zeroes.
```

extensions, OuterExtensions, includes those which were removed from
EncodedClientHelloInner, in the order in which they were removed.

Finally, the client sets the `padding` field to a byte string whose contents
are all zeros. {{padding}} describes a recommended padding scheme.

The client-facing server computes ClientHelloInner by reversing this process.
First it parses EncodedClientHelloInner, interpreting all bytes after

> `client_hello` as padding. Next it makes a copy of the `client_hello` field and

Reviewer 3: Should we require the server check that padding is all zeroes?

Author: Sure. Done.

extensions, OuterExtensions, includes those which were removed from
EncodedClientHelloInner, in the order in which they were removed.

Finally, the client sets the `padding` field to a byte string whose contents

> are all zeros. {{padding}} describes a recommended padding scheme.

Reviewer: We ought to define N in this paragraph. Something like: Finally, the client sets the `padding` field to a length-N all-zero byte string, where N is computed as described in {{padding}}.`

Author: Done, although tweaked slightly since {{padding}} is merely recommended, not required.

> Finally, the client sets the `padding` field to a byte string whose contents
> are all zeros. {{padding}} describes a recommended padding scheme.

Reviewer 3: ```suggestion
Finally, the client sets the `padding` field to a byte string of length N whose
contents are all zeros. {{padding}} describes a recommended padding scheme.
```

Author: Done, although tweaked slightly since {{padding}} is merely recommended, not required.

extensions, OuterExtensions, includes those which were removed from
EncodedClientHelloInner, in the order in which they were removed.

> Finally, the client determines the number of padding bytes, N, and sets the

Reviewer: nit (here and below):
```suggestion
Finally, the client determines the number of padding bytes, length_of_padding, and sets the
```

Author: Well, N is referenced immediately afterwards and elsewhere we've used capital letters for random variables in prose. Note RFC8446 never utters length_of_padding in prose either. (Really what's going on here is that this is a questionably valid TLS message. The TLS message syntax is typically self-delimiting. But it's silly to add a length prefix here.)

I'll rewrite this to not reference a variable name.

ClientHello extensions can be computed in this way.

In contrast, clients do not know the longest SNI value in the client-facing
server's anonymity set without server input. Clients SHOULD use the ECHConfig's
`maximum_name_length` field as follows, where L is the `maximum_name_length`
value.

1. If the ClientHelloInner contained a "server_name" extension with a name of
   length D, add max(0, L - D) bytes of padding.

> 2. If the ClientHelloInner did not contain a "server_name" extension (e.g. if

Reviewer 2: ```suggestion
2. If the ClientHelloInner did not contain a "server_name" extension (e.g., if
```

ClientHello extensions can be computed in this way.

In contrast, clients do not know the longest SNI value in the client-facing
server's anonymity set without server input. Clients SHOULD use the ECHConfig's
`maximum_name_length` field as follows, where L is the `maximum_name_length`
value.

1. If the ClientHelloInner contained a "server_name" extension with a name of
   length D, add max(0, L - D) bytes of padding.
2. If the ClientHelloInner did not contain a "server_name" extension (e.g., if
   the client is connecting to an IP address), add L + 7 bytes of padding. This
   is the length of a "server_name" extension with an L-byte name.

Finally, the client SHOULD pad the entire message as follows:

1. Let L be the length of the EncodedClientHelloInner with all the padding
   computed so far.
2. Let P = 31 - ((L - 1) % 32) and add P bytes of padding.

> This rounds the overall structure up to a multiple of 32 bytes. This reduces
> the range of lengths of the EncodedClientHelloInner across all clients.

Reviewer 3: ```suggestion
This rounds the length of EncodedClientHelloInner up to a multiple of 32 bytes,
reducing lengths of this structure across all clients.
```

Author: Agreed on the first half. I worry "reducing lengths" reads like making the lengths smaller. I guess "range of lengths" also has a different meaning... I was thinking the range of a function. :-)

How about "reducing the set of possible lengths across all clients".

Reviewer 3: That works for me too 👍

Author: Done.

ClientHello extensions can be computed in this way.

In contrast, clients do not know the longest SNI value in the client-facing
server's anonymity set without server input. Clients SHOULD use the ECHConfig's
`maximum_name_length` field as follows, where L is the `maximum_name_length`
value.

1. If the ClientHelloInner contained a "server_name" extension with a name of
   length D, add max(0, L - D) bytes of padding.
2. If the ClientHelloInner did not contain a "server_name" extension (e.g., if
   the client is connecting to an IP address), add L + 9 bytes of padding. This
   is the length of a "server_name" extension with an L-byte name.

Finally, the client SHOULD pad the entire message as follows:

1. Let L be the length of the EncodedClientHelloInner with all the padding
   computed so far.

> 2. Let P = 31 - ((L - 1) % 32) and add P bytes of padding.

Reviewer: s/P/N/g

Author: Done.

> 2. Let P = 31 - ((L - 1) % 32) and add P bytes of padding.

Reviewer 3: ```suggestion
2. Let N = 31 - ((L - 1) % 32) and add N bytes of padding.
```

Author: Done.
